### DISCUSSION TITLE
进度检测监控疑问

### DISCUSSION THREAD

**Discussion by 135ty** [PBH-BTN/PeerBanHelper#590] [问答] [answered]
我在下载一个种子的时候假设进度下载到8.5%的时候连接不上做种，我用第三方软件离线下载后替换了这个下载文件，重新校验后进度变成100%，进入做种模式。假若这个行为被其他开着PBH监控的bt客户端用户检测到，是否会被当作作弊行为被封禁？如果进度是50%的时候这样做也会被封禁吗？

**Comment by Ghost-chu** [comment, marked as answer]
> 重新校验后进度变成100%，进入做种模式。假若这个行为被其他开着PBH监控的bt客户端用户检测到，是否会被当作作弊行为被封禁？如果进度是50%的时候这样做也会被封禁吗？

不会，PBH 只管变少不管变多。但一旦变多就不能再变少了。

> 我在下载一个种子的时候假设进度下载到8.5%的时候连接不上做种，我用第三方软件离线下载后替换了这个下载文件

如果你说的第三方软件在 PeerID/ClientName 的黑名单中，并且连接到了受 PBH 保护的用户，会导致你的 IP 被封禁。封禁期间，即使你更换回了正常下载器，也无法再解封，只能等待封禁过期。  
不过，由此被封禁的 IP 仅在对应用户的 PBH 上生效，不会与其它人通过 BTN 网络共享。也不参与规则生成。

**Reply by fejeIaqbhbv** [reply]
你好，我也想问一下

如果同一个种子原本已经下载完并持续做种，某天不小心删了，立刻校验重新下载也会进名单吗？

**Reply by Ghost-chu** [reply]
> 你好，我也想问一下
> 
> 如果同一个种子原本已经下载完并持续做种，某天不小心删了，立刻校验重新下载也会进名单吗？

取决于你连接的人中是否包含 PBH 用户，如果没有则不会；如果有的话。取决于它们上次看到过你的时间。如果超过了持久化存储时间，你就是新 Peer。反之则会触发封禁。

对于是否会被加入 BTN 在线规则，这取决于多少 PBH 决定封禁你的 IP。一旦被列入在线规则，则需要等待最少 45 天的时间才能被移除。

对于普通家宽用户，一般不需要有这种担忧。因为你把光猫路由器重启一下，IP自己就换了。

**Reply by fejeIaqbhbv** [reply]
> > 如果同一个种子原本已经下载完并持续做种，某天不小心删了，立刻校验重新下载也会进名单吗？
> 
> 取决于你连接的人中是否包含 PBH 用户，如果没有则不会；如果有的话。取决于它们上次看到过你的时间。如果超过了持久化存储时间，你就是新 Peer。反之则会触发封禁。
> 对于是否会被加入 BTN 在线规则，这取决于多少 PBH 决定封禁你的 IP。一旦被列入在线规则，则需要等待最少 45 天的时间才能被移除。

你好开发者，我想了一下

我认为应该把声明100%进度从检测机制中移出去，这样误封的概率很大。只检测小于100%的

以您的说法，持续做种就表示一直存在于持久化存储的档案中。
就是说只要手误删文件，就直接进黑名单
重启客户端时，做种目录丢失，如果没立即重新修正目录，声明进度也会降为0%，同样进黑名单

我搜了一下相关信息。BT协议不支持声明100%的同时还可以索要文件。也没有找到有已经声明100%的客户端还会向其它客户端索要文件的相关内容

所以存储的应当是声明最后一次小于100%的进度信息，不包含等于100%的
这样即使进度丢失，由于做种时间已经超过持久化存储时间，上一份小于100%的进度信息已经过期，就是新的peer

> 对于普通家宽用户，一般不需要有这种担忧。因为你把光猫路由器重启一下，IP自己就换了。

我对这句话不是很认同，既然现在有连坐机制，应该更加小心以防止触发连坐。

**Reply by Ghost-chu** [reply]
> 我搜了一下相关信息。BT协议不支持声明100%的同时还可以索要文件。也没有找到有已经声明100%的客户端还会向其它客户端索要文件的相关内容

协议没说 100% 还可以索要文件，但也没说 100% 就不能索要文件。发送精心构造的请求包后，可强制转换下载器，特别是 qBittorrent，来向恶意客户端全速发送上传分片。在我们的内部的实验室中已经成功成功在多位组员的下载器上成功复现了此问题。

**Reply by Ghost-chu** [reply]
> 重启客户端时，做种目录丢失，如果没立即重新修正目录，声明进度也会降为0%，同样进黑名单

做种文件丢失时，所有主流下载器均会停止任务，并置于错误状态。此时下载器不会主动外连或接受与此种子有关的传入连接。

除非您在看到错误后，依旧选择了重新下载，开始消耗所必须之外的流量，那么你的行为就已经和吸血客户端一样了。被封禁没有问题。

**Reply by fejeIaqbhbv** [reply]
那如果是  一个周期内(比如上文说的45天)允许一次从100%降到任意数值  的行为呢？

客户端发现是从100%降的，就只记录ip不封禁，BTN核对只有1个或2个种子，就只存档，周期内再发现就下发封禁ip。
可以上升至IP段连坐，比如一个/24底下一个周期只允许存在10个，第11个发现就直接封禁。

虽然我知道这种行为不是一个好行为，但也不是恶意行为，我希望能留一个容错。

**Reply by GitHubUserNameTwo** [reply]
我也差不多，原种子下载到百分之八十没资源了，找了五六个具有相同文件的种子继续下载，糟糕的是校检原文件没有反应，点了继续下载结果把原文件覆盖清空的，只能全部重新下载。又觉得成功下载的希望不大，最后借助云盘下载好把文件补进去了，在别人看来我的进度肯定也是在0到100反复横跳。当然我觉得这种情况被误封也无可奈何，如果有同一用户的综合行为分析就更完美了，应该很难做到吧

**Reply by Ghost-chu** [reply]
> 我也差不多，原种子下载到百分之八十没资源了，找了五六个具有相同文件的种子继续下载，糟糕的是校检原文件没有反应，点了继续下载结果把原文件覆盖清空的，只能全部重新下载。又觉得成功下载的希望不大，最后借助云盘下载好把文件补进去了，在别人看来我的进度肯定也是在0到100反复横跳。当然我觉得这种情况被误封也无可奈何，如果有同一用户的综合行为分析就更完美了，应该很难做到吧

很难做到，但一般来说不会受到严重影响。对于家宽用户来说，即使被封禁了也可以重启路由器更换 IP 来解决。  
但如果你所在区县大量这种类似的记录，则可能导致附近的 IP 段被列入黑名单。

**Reply by CalunVier** [reply]
> 不会，PBH 只管变少不管变多。但一旦变多就不能再变少了。

关于这句话。如果我一开始只下载了种子内的部分内容。下载进度到达100%之后（或者没到），我又选中下载种子内的其他内容。此时进度自然而然会产生回退。请问这种情况是否会被当作作弊行为？

更复杂的情况。如果我下载完种子中的文件A后，移除文件A，同时取消勾选下载文件A；然后勾选并开始下载文件B。这种行为是否会被当作作弊行为？

我注意到 7.2.0 更新公告 #806 内有相关内容

> 改进了对 [BEP-0021 Extension for partial seeds](https://www.bittorrent.org/beps/bep_0021.html) 规范的支持，现在能够对下列情况做出正确的过量下载/进度回退的反吸血判定：
> 
> 种子处于下载状态，部分文件没有完整下载
> 用户/Peer在 “内容” 选项卡中只选择了部分文件下载，而不是整个完整的 torrent (BEP-0021)
> 下载过程中出现错误，导致下载停止，进入了仅上传状态（如 I/O 错误，或者在下载 torrent 过程中磁盘空间不足）

这是否影响到了 #621 中提及的
> 你改变下载的文件会对进度有影响，只要进度是倒退的检测到就会被ban，不会考虑到改文件的情况。
？

**Reply by Ghost-chu** [reply]
> > 不会，PBH 只管变少不管变多。但一旦变多就不能再变少了。
> 
> 关于这句话。如果我一开始只下载了种子内的部分内容。下载进度到达100%之后（或者没到），我又选中下载种子内的其他内容。此时进度自然而然会产生回退。请问这种情况是否会被当作作弊行为？
> 
> 更复杂的情况。如果我下载完种子中的文件A后，移除文件A，同时取消勾选下载文件A；然后勾选并开始下载文件B。这种行为是否会被当作作弊行为？

这取决于最终客户端的实现，如果从他人角度看到你的客户端进度发生回退，它就会被计入。  

不同客户端行为不同。

**Reply by CalunVier** [reply]
抱歉，我在修改上一条提问的时候没有看到您的回复。

我注意到 7.2.0 更新公告 https://github.com/PBH-BTN/PeerBanHelper/discussions/806 内有相关内容


> 改进了对 [BEP-0021 Extension for partial seeds](https://www.bittorrent.org/beps/bep_0021.html) 规范的支持，现在能够对下列情况做出正确的过量下载/进度回退的反吸血判定：
> 
> 
> 种子处于下载状态，部分文件没有完整下载
> 用户/Peer在 “内容” 选项卡中只选择了部分文件下载，而不是整个完整的 torrent (BEP-0021)
> 下载过程中出现错误，导致下载停止，进入了仅上传状态（如 I/O 错误，或者在下载 torrent 过程中磁盘空间不足）

这是否影响到了 https://github.com/PBH-BTN/PeerBanHelper/discussions/621 中提及的

> 你改变下载的文件会对进度有影响，只要进度是倒退的检测到就会被ban，不会考虑到改文件的情况。

？如果行为没有改变，我该如何正确理解关于 BEP-0021 适配性的改进？

此外，我使用 BiglyBT 和 qBittorrent Enhancement Editon 作为我的主要下载器，请问这两个的实现行为如何？
